### DISCUSSION TITLE
مشکل در  پاسخ از بازگشت  در گاه تست

### DISCUSSION THREAD

**Discussion by Sajol003** [Sina-Soltani/Parbad#313] [Q&A]
با سلام  من درگاه تست رو پیاده سازی کردم  کاربر به سمت درگاه هدایت میشه و مشکلی نداره و لی بازگشت و پاسخ از درگاه  مشکل دارم 
متد Verify  در بازگشت فراخوانی نمیشه  فکر  میکنم در callback url مشکل دارم  پروژه ام razorpages  مقادیر دریافتی  url.action  نام کنترلر رو نمیدونم چی باید وارد کنم 
نمنونه کدم رو قرار میدم ممنون میشم راهنماییم کنید 
payment ()

                  var callbackUrl = Url.Action("Verify", "Payment", null, Request.Scheme);
                var result = await _onlinePayment.RequestAsync(invoice =>
                {
                    invoice
                       .SetTrackingNumber(123)
                       .SetAmount(factor.Amount)
                       .SetCallbackUrl(callbackUrl)
                       .SetGateway("ParbadVirtual")
                       // OR
                       .UseParbadVirtual();
                       });

                if (result.IsSucceed)
                {
                    return result.GatewayTransporter.TransportToGateway();
                }

[HttpGet, HttpPost]
        public async Task<IActionResult> Verify(int id)
        {
        
            var invoice = await _onlinePayment.FetchAsync();

            
            if (invoice.Status == PaymentFetchResultStatus.AlreadyProcessed)
            {
                
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                return Content("The payment is already processed before.");
            }

            
          

            var verifyResult = await _onlinePayment.VerifyAsync(invoice);
            if (verifyResult.IsSucceed)
            {
        
                 return Page();
                Factor.Ispay = true;
            }
           

          

            return Page();
        }

**Comment by AliDeV7** [comment]
سلام، مشکل همون قسمت Callback Url هست که شما کامل پیاده سازی نکردی

``` 
var HostBaseUrl = $"{contextAccessor.HttpContext.Request.Scheme}://{contextAccessor.HttpContext.Request.Host}{contextAccessor.HttpContext.Request.PathBase};

String CallbackURL = HostBaseUrl +"/Payment/Verify";
```
یک سری دیتا هم (فکر میکنم حتما لازم داری وقتی از بانک به برنامه برمیگردی) باید اضافه کنی که میتونی به URL به صورت QueryString اضافه کنی

**Comment by Sajol003** [comment]
با سلام .درود   بعد از پرداخت در درگاه بانک  به صفحه مورد نظر هدایت میشم  و بانک به من یک PaymentToken میده  ولی صفحه مورد نظر تو بازگشت اول فراخوانی نمیشه !!! وقتی دباره همون url رو Reload میکنم اینبار صفحه مورد نظر فراخوانی میشه  
خط   var invoice = await _onlinepayment.FetchAsync(); به درستی کار میکنه  ولی خط 
  var verifyResult = await _onlinepayment.VerifyAsync(invoice);
نتیجه درست رو بر نمی گردونه  در هر صورتی False بر میگردونه

**Reply by AliDeV7** [reply]
شما وقتی که Reload میکنی، مشخصه که به ارور میخوری، چون درخواست فقط یکبار کار میکنه و درست نیست که با Reload هم دوباره True برگردونه

**Comment by AliDeV7** [comment]
کدوم درگاه بانک رو الان دارین تست میکنین؟

**Reply by Sajol003** [reply]
دارم رو درگاه تست parbad  کار میکنم   
بازم یجای کارم میلنگه 
به قسمت stsartup رفتم  و در بخش تنظیمات پرباد خطی وجود دارد  که  gatwaypath  رو میدیم  
.WithOptions(options => options.GatewayPath = "/OnlinePayProduct");
 
تا قبل از این  مقدار gatwaypath رو دست کاری نکرده بودم و همون مقدار پیش فرض مستدات بود  اون رو به  OnlinePayProduct تغیر دادم که صفحه 
دریافت پاسخ از بانک  هست 
این بار با پیغامی مواجه شدم (Command details are not valid.) نمیدونم کجای کارم اشتباست 
سورس کامل پیاده سازیمو میزارم ممنون میشم بررسی کنید و راهنماییم کنید 

startup   استارت آپ
builder.Services.AddParbad()
.ConfigureGateways(gateways =>
                {
                    gateways
                        .AddParbadVirtual()
                        .WithOptions(options => options.GatewayPath = "/OnlinePayProduct");
                })
.ConfigureHttpContext(builder => builder.UseDefaultAspNetCore())
.ConfigureStorage(builder => builder.AddStorage<MyStorage>(ServiceLifetime.Transient));



Payment  پرداخت

var HostBaseUrl = $"{HttpContext.Request.Scheme}://{HttpContext.Request.Host}{HttpContext.Request.PathBase}";
string callback = HostBaseUrl + "/OnlinePayProduct?id=" + factor.id;
 var result = await _onlinePayment.RequestAsync(invoice =>
                {
                    invoice
                       .SetTrackingNumber(123)
                       .SetAmount(factor.Amount)
                       .SetCallbackUrl(callback)
                       .SetGateway("ParbadVirtual")
                       // OR
                       .UseParbadVirtual();
                       });


                if (result.IsSucceed)
                {
                    return result.GatewayTransporter.TransportToGateway();
                }

OnlinepayProduct    دریافت پاسخ از بانک 
 public async Task<IActionResult> OnGet(int id)
        {
            var Factor = _contex.factors.Find(id);
            var invoice = await _onlinepayment.FetchAsync();

   
            if (invoice.Status == PaymentFetchResultStatus.AlreadyProcessed)
            {
           
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                return Content("The payment is already processed before.");
            }

            
            if (!Is_There_Still_Product_In_Shop(invoice.TrackingNumber))
            {
                var cancelResult = await _onlinepayment.CancelAsync(invoice, cancellationReason: "Sorry, We have no more products to sell.");

                //return View("CancelResult", cancelResult);
            }

            var verifyResult = await _onlinepayment.VerifyAsync(invoice);
            if (verifyResult.IsSucceed)
         


            Factor.Ispay = true;
        }
            return Page();

**Comment by Sina-Soltani** [comment]
درود
من اصلا متوجه مشکلاتی که اینجا عنوان کردید نمیشم چون اول گفتید که متد وریفای فراخوانی نمیشه، بعد گفتید که میشه ولی مقدار فالس برمیگردونه، و بعد میگید شاید مشکل از آدرس بازگشتی باشه.

مشکل رو یک بار دیگه به روش صحیح و واضح تر توضیح بدید تا در رفع مشکل بهتون کمک لازم رو انجام بدم.
قدم به قدم از مرحله اول که ریکوئست هست توضیح بدید تا جایی که مشکل به وجود میاد.

**Reply by Sajol003** [reply]
درود 
من تو مرحله وصل شدن به درگاه مشکلی ندارم اون مراحل به درستی انجام میشه ولی موقع بازگشت از درگاه  به صفحه  مورد نظر (callbackurl) هدایت میشم  ولی صفحه تو مرتبه اول کار نمیکنه یعنی برک پوینت که قرار دادم اصلا بهش نمیرسه در صورتی url که به سمتش هدایت شدم کاملا درسته حالا برای بار دوم  url رو reload میکنم  صفحه مورد نظر کار میکنه ولی  نتیجه پرداخت در هر شکلی false تصور خودم اینه که callbackurl رو به فرمتی درست وارد نمیکنم نمونه کد رو قرار میدم ممنون میشم برسی کنید  و راهنماییم کنید 


public async Task<IActionResult> Payment(int id)
{
var HostBaseUrl = $"{HttpContext.Request.Scheme}://{HttpContext.Request.Host}{HttpContext.Request.PathBase}";

string callback = HostBaseUrl + "/OnlinePayProduct?id=" + factor.id;

var result = await _onlinePayment.RequestAsync(invoice =>
{
invoice
.SetTrackingNumber(123)
.SetAmount(factor.Amount)
.SetCallbackUrl(callback)
.SetGateway("ParbadVirtual")

.UseParbadVirtual();
});


            if (result.IsSucceed)
            {
                return result.GatewayTransporter.TransportToGateway();
            }
}





/OnlinePayProduct


public async Task OnGet(int id)
{
var Factor = _contex.factors.Find(id);
var invoice = await _onlinepayment.FetchAsync();

        if (invoice.Status == PaymentFetchResultStatus.AlreadyProcessed)
        {
       
            var isAlreadyVerified = invoice.IsAlreadyVerified;
            return Content("The payment is already processed before.");
        }

        
        if (!Is_There_Still_Product_In_Shop(invoice.TrackingNumber))
        {
            var cancelResult = await _onlinepayment.CancelAsync(invoice, cancellationReason: "Sorry, We have no more products to sell.");

            //return View("CancelResult", cancelResult);
        }

        var verifyResult = await _onlinepayment.VerifyAsync(invoice);
        if (verifyResult.IsSucceed)
     


        Factor.Ispay = true;
    }
        return Page();

**Comment by Sina-Soltani** [comment]
این مسائل رو شما باید در گوگل یا فروم‌های برنامه نویسی دنبال کنید نه اینجا چون نحوه تولید یک آدرس در دات نت، واقعا هیچ ارتباطی با این ابزار نداره.

مقدار
callbackUrl
 آدرسی در وبسایت شما هست که در اون قسمت شما کدهای تایید تراکنش یا به اصطلاح همون متد وریفای رو نوشتید.

این ارتباطی اصلا به درگاه تست یا هیچ درگاه دیگه ای نداره. این یک آدرس در وب سایت شما هست که وقتی خریدار از درگاه بانک مثلا سامان برمیگرده به وب سایت شما، در اون آدرس شما کدهای تایید پرداخت رو باید نوشته باشید.

اگر کدهای قسمت تایید پرداخت یعنی همون متد وریفای رو برای مثال در آدرسی مانند آدرس زیر دارید:
http://yourwebsite.com/payment/verify
بنابراین آدرس callbackUrl هم برابر میشه با همین آدرس.

در صورتیکه شما این مقدار رو برابر با آدرس درگاه تست که در مرحله تنظیمات هست قرار دادید که در واقع دو تا چیز کاملا بی ربط هستن

**Reply by Sajol003** [reply]
بله خیلی ممنون از پاسخگوییتون

**Reply by Sina-Soltani** [reply]
خواهش میکنم

**Reply by Dev-AMF** [reply]
سلام و خسته نباشید
بزارید مشکل دوستمون رو که منم اخیرا بهش برخوردم رو توضیح بدم تا شاید واضح تر بشه صورت مسئله.
شما درگاه مجازی رو طبق داکیومنت ها در فایل startup کانفیگ میکنی.
بعداز اونجایی که پروژه ت از نوع RazorPages هست، میری توی Page Handler ت
و یک ریکوئست ایجاد میکنی تا کاربر به درگاه مجازی منتقل بشه و در صورت پرداخت موفق آمیز با status ئه True به CallbackUrl برگرده.
اما با اینکه شما CallbackUrl صحیحی تولید کردی ولی در کمال ناباوری میبینی که ارور 400 میگیری!
حالا از کجا متوجه میشی که CallbackUrl شما صحیحه؟
وقتی url ای که درگاه کاربر رو بهش منتقل میکنه رو کپی پیست میکنی توی یه تب جدید و اینتر رو میزنی برک پوینت شما هیت میشه و کاربر دقیقا میره همونجایی که باید میرفت. خب پس مشکل کجاست؟Staus پرداخت درحالی که باید True باشه Fasle هست که طبق [گفته دوستمون](https://github.com/Sina-Soltani/Parbad/discussions/313#discussioncomment-3055621) بخاطر Reload شدن هست و اونم به صورت دستی!
حالا من یه سری اسکرین شات که شرایط فوق رو نشون میده قرار میدم و همچنین کدهایی که لازمه رو شاید به درک بهتر موضوع کمک کرد و ممنون میشم اگر کسی میتونه در این باره مارو راهنمایی کنه چه اشکال از ما باشه چه از درگاه 🙏

![NewTab](https://user-images.githubusercontent.com/113021373/201518340-017ff1ce-8510-4bde-b52c-af0208555f1b.PNG)
![NotFoundPage](https://user-images.githubusercontent.com/113021373/201518402-5af777d4-4b95-4a78-9ec3-172e4ef05aa6.PNG)
![UnSuccessfulPayment](https://user-images.githubusercontent.com/113021373/201518404-9ebdc194-22f5-478e-b3b4-9a8778fb7e1c.PNG)
![BreakpointHeat](https://user-images.githubusercontent.com/113021373/201518405-6a698829-fe0c-4ff3-82f3-0debffe92ac4.PNG)
![Versions](https://user-images.githubusercontent.com/113021373/201520165-ee9e0f67-c350-4fb3-af66-9f5f3c84b4bb.PNG)
-----------------------------------------------------------------------------------------------------------------------------------------------------
کد های زیر به ترتیب برای دو Page Handler ای که گفتم و کد زیرش هم کد Config کردن درگاه در فایل Startup ست.
----------------------------------------------------------------------------------------------------------------------------------------------------


`public IActionResult OnGetPay()
        {
            var cart = _cartService.Get();
            var result = _productQuery.CheckInventoryStatus(cart.Items);

            if (result.Any(CI => CI.IsInStock) == false)
                return RedirectToPage("/Cart");

            var orderId = _orderApp.PlaceOrder(cart);

            var siteUrl = _configuration.GetSection("payment")["siteUrl"];
            var CallbackURL = $"{siteUrl}/Checkout?handler=CallBack&oId={orderId}";

            var payresult = _onlinePayment.Request(invoice =>
                               invoice
                               .UseAutoRandomTrackingNumber()
                               .SetCallbackUrl(CallbackURL)
                               .SetAmount((Money)cart.PayAmount)
                               .SetGateway("ParbadVirtualGateway")
                               .UseParbadVirtual()

            );

               GatewayTrackNo = payresult.TrackingNumber.ToString();

                var url =  payresult.GatewayTransporter.TransportToGateway();
            
            return payresult.GatewayTransporter.TransportToGateway();
        }

        public IActionResult OnGetCallBack([FromQuery] int oId, [FromQuery] string paymentToken)
        {
            var result = new PaymentResult();

            var verificationResponse = _onlinePayment.Verify(long.Parse(GatewayTrackNo));
            
            var RefId = verificationResponse.TrackingNumber;
            var TrackingNo = verificationResponse.TransactionCode;

            if (verificationResponse.IsSucceed)
            {
                var issueTrackingNo = _orderApp.PaymentSucceeded(oId, RefId , TrackingNo);
                
                Response.Cookies.Delete("cart-items");
                result = result.Succeeded(ApplicationMessages.SuccessfulPayment, issueTrackingNo);
                
                return RedirectToPage("/PaymentResult", result);
            }
            else
            {
                Response.Cookies.Delete("cart-items");

                result = result.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", result);
            }
        }`        
`            services.AddParbad()
                .ConfigureGateways(gateways =>
                {
                    gateways
                         .AddMellat()
                         .WithAccounts(source =>
                         {
                             source.AddInMemory(account =>
                             {
                                 account.TerminalId = 123;
                                 account.UserName = "MyId";
                                 account.UserPassword = "MyPassword";
                             });
                         });

                    gateways.AddParbadVirtual()
                            .WithOptions(options => options.GatewayPath = "/MyvirtualGateway");

                })
                .ConfigureHttpContext(builder => builder.UseDefaultAspNetCore())
                .ConfigureStorage(builder => builder.UseMemoryCache())
                .ConfigureAutoRandomTrackingNumber(options =>
                 {
                     options.MinimumValue = 1000;
                 });

**Comment by Sina-Soltani** [comment]
درود
فکر میکنم علت اینکه شما ۴۰۴ دریافت می کنید اینه که متد کال بک شما فقط حالت GET رو داره ساپورت میکنه، در صورتیکه درگاه مجازی و همچنین تعدادی از درگاه‌های بانکی، اطلاعات رو در حالت POST ارسال میکنند. متد شما باید هر دو حالت رو پشتیبانی کنه.
در نمونه مثال‌ها، این مورد دقیقا ذکر شده
https://github.com/Sina-Soltani/Parbad.Samples/blob/master/Web/Parbad.Sample.AspNetCore/Controllers/PaymentController.cs#L55

همچنین تلاش کنید ساخت url رو به این صورت به حالت دستی انجام ندید و از امکانات دات نت استفاده کنید.
```c#
var callbackUrl = Url.Page("pageName", "handlerName", extraRouteValues, Request.Scheme);
```

در ضمن اگر مستندات و نمونه کد‌ها رو مشاهده کنید،‌می بینید که از متد Fetch قبل از وریفای استفاده شده.

**Reply by Dev-AMF** [reply]
درود مجدد
من برای متد کال بک خودم هندلر Post هم تعریف کردم و الان هم حالت Post  رو ساپورت میکنه هم حالت Get
( در RazorPages امکان استفاده از [HttpGet, HttpPost] بر روی هندلرها نیست و یک هندلر به طور همزمان فقط یک حالت رو ساپورت میکنه به همین دلیل باید یک هندلر همنام برای متد دیگر نوشته شه )
متد Fetch رو هم پیاده سازی کردم و برای ساخت Url هم به شکل گفته شده عمل کردم،  حتی حالت Async رو هم تست کردم ولی همچنان :

![ParbadCallBack Error](https://user-images.githubusercontent.com/113021373/202448777-ba165eb5-0845-48e2-9b55-6c5ff4fbb3a3.png)
نمونه کد ها: 


`        public IActionResult OnPostPay()
        {
            var cart = _cartService.Get();
            var result = _productQuery.CheckInventoryStatus(cart.Items);

            if (result.Any(CI => CI.IsInStock) == false)
                return RedirectToPage("/Cart");

            var orderId = _orderApp.PlaceOrder(cart);

            OrderId = orderId.ToString();
            
            var CallbackURL = Url.Page("./Checkout", "Callback", null, Request.Scheme, Request.Host.Value);


            var payresult =  
                _onlinePayment.Request(invoice =>
                               invoice
                               .UseAutoRandomTrackingNumber()
                               .SetCallbackUrl(CallbackURL)
                               .SetAmount((Money)cart.PayAmount)
                               .SetGateway("ParbadVirtualGateway")
                               .UseParbadVirtual()
            );

            GatewayTrackNo = payresult.TrackingNumber.ToString();

            var url = payresult.GatewayTransporter.TransportToGateway();

            return payresult.GatewayTransporter.TransportToGateway();
        }

        public  IActionResult OnGetCallBack()
        {
            var payresult = new PaymentResult();
            
            var invoice =  _onlinePayment.Fetch();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status != PaymentFetchResultStatus.ReadyForVerifying)
            {
                // You can also see if the invoice is already verified before.
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }


            var verificationResponse =  _onlinePayment.Verify(long.Parse(GatewayTrackNo));

            var RefId = verificationResponse.TrackingNumber;
            var TrackingNo = verificationResponse.TransactionCode;
            var oId = int.Parse(OrderId);

            if (verificationResponse.IsSucceed)
            {
                var issueTrackingNo = _orderApp.PaymentSucceeded(oId, RefId, TrackingNo);

                Response.Cookies.Delete("cart-items");
                payresult = payresult.Succeeded(ApplicationMessages.SuccessfulPayment, issueTrackingNo);

                return RedirectToPage("/PaymentResult", payresult);
            }
            else
            {
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }
        }
        public IActionResult OnPostCallBack()
        {
            var payresult = new PaymentResult();

            var invoice =  _onlinePayment.Fetch();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status != PaymentFetchResultStatus.ReadyForVerifying)
            {
                // You can also see if the invoice is already verified before.
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }

            var verificationResponse =  _onlinePayment.Verify(long.Parse(GatewayTrackNo));

            var RefId = verificationResponse.TrackingNumber;
            var TrackingNo = verificationResponse.TransactionCode;
            var oId = int.Parse(OrderId);

            if (verificationResponse.IsSucceed)
            {
                var issueTrackingNo = _orderApp.PaymentSucceeded(oId, RefId, TrackingNo);

                Response.Cookies.Delete("cart-items");
                payresult = payresult.Succeeded(ApplicationMessages.SuccessfulPayment, issueTrackingNo);

                return RedirectToPage("/PaymentResult", payresult);
            }
            else
            {
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }
        }`


حالت Async :




`public async Task<IActionResult> OnPostPay()
        {
            var cart = _cartService.Get();
            var result = _productQuery.CheckInventoryStatus(cart.Items);

            if (result.Any(CI => CI.IsInStock) == false)
                return RedirectToPage("/Cart");

            var orderId = _orderApp.PlaceOrder(cart);

            OrderId = orderId.ToString();
            
            var CallbackURL = Url.Page("./Checkout", "Callback", null, Request.Scheme, Request.Host.Value);


            var payresult = await
                _onlinePayment.RequestAsync(invoice =>
                               invoice
                               .UseAutoRandomTrackingNumber()
                               .SetCallbackUrl(CallbackURL)
                               .SetAmount((Money)cart.PayAmount)
                               .SetGateway("ParbadVirtualGateway")
                               .UseParbadVirtual()
            );

            GatewayTrackNo = payresult.TrackingNumber.ToString();

            var url = payresult.GatewayTransporter.TransportToGateway();

            return payresult.GatewayTransporter.TransportToGateway();
        }

        public async Task<IActionResult> OnGetCallBack()
        {
            var payresult = new PaymentResult();
            
            var invoice = await _onlinePayment.FetchAsync();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status != PaymentFetchResultStatus.ReadyForVerifying)
            {
                // You can also see if the invoice is already verified before.
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }


            var verificationResponse = await _onlinePayment.VerifyAsync(long.Parse(GatewayTrackNo));

            var RefId = verificationResponse.TrackingNumber;
            var TrackingNo = verificationResponse.TransactionCode;
            var oId = int.Parse(OrderId);

            if (verificationResponse.IsSucceed)
            {
                var issueTrackingNo = _orderApp.PaymentSucceeded(oId, RefId, TrackingNo);

                Response.Cookies.Delete("cart-items");
                payresult = payresult.Succeeded(ApplicationMessages.SuccessfulPayment, issueTrackingNo);

                return RedirectToPage("/PaymentResult", payresult);
            }
            else
            {
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }
        }
        public async Task<IActionResult> OnPostCallBack()
        {
            var payresult = new PaymentResult();

            var invoice = await _onlinePayment.FetchAsync();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status != PaymentFetchResultStatus.ReadyForVerifying)
            {
                // You can also see if the invoice is already verified before.
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }

            var verificationResponse = await _onlinePayment.VerifyAsync(long.Parse(GatewayTrackNo));

            var RefId = verificationResponse.TrackingNumber;
            var TrackingNo = verificationResponse.TransactionCode;
            var oId = int.Parse(OrderId);

            if (verificationResponse.IsSucceed)
            {
                var issueTrackingNo = _orderApp.PaymentSucceeded(oId, RefId, TrackingNo);

                Response.Cookies.Delete("cart-items");
                payresult = payresult.Succeeded(ApplicationMessages.SuccessfulPayment, issueTrackingNo);

                return RedirectToPage("/PaymentResult", payresult);
            }
            else
            {
                Response.Cookies.Delete("cart-items");

                payresult = payresult.Failed(ApplicationMessages.UnSuccessfulPayment);
                return RedirectToPage("/PaymentResult", payresult);
            }
        }`

**Reply by Sina-Soltani** [reply]
درود
۴۰۴ دریافت کردید یا چیز دیگه ای؟
اگر ۴۰۴ دریافت کردید، بعد از مشاهده این صفحه، یک یا چند بریک پوینت بزارید روی کد‌ها و بعد روی نوار آدرس در مرورگر کلیک کنید و اینتر بزنید تا ببینید وارد کد‌های شما میشه یا خیر.
چیزی که در واقع باید تست کنید آیا همین صفحه با همین آدرس توی حالت GET وارد کدها میشه یا خیر
